File: cgi-bin/cmd.py
#!/usr/bin/python
#  This file is part of Leash (Browser Shell)
#  Copyright (C) 2013-2014  Jakub Jankiewicz <http://jcubic.pl>
#
#  Released under the MIT license

import cgitb; cgitb.enable()
import subprocess, os
from sys import stdout, stdin
import json

def shell_exec(code):
    try:
        return subprocess.check_output(code, shell=True)
    except subprocess.CalledProcessError as e:
        return e.output

def trace():
    import traceback, StringIO
    buff = StringIO.StringIO()
    traceback.print_exc(file=buff)
    return buff.getvalue()

if __name__ == '__main__':
    print "Content-Type: application/json"
    print
    import os
    from utils import valid_token
    response = {}
    if not os.environ['REMOTE_ADDR'] == os.environ['SERVER_ADDR']:
        response['error'] = 'You can access this script only from same server ' + \
                            '(Service.php script)'
    elif valid_token():
        try:
            response['result'] = shell_exec(stdin.read())
        except Exception, e:
            response['error'] = e.args[0]
            response['trace'] = trace()
    else:
        response['error'] = "The token is invalid"
    stdout.write(json.dumps(response))

File: cgi-bin/cmd.pl
#!/usr/bin/perl -w
#  This file is part of Leash (Browser Shell)
#  Copyright (C) 2013-2018  Jakub Jankiewicz <http://jcubic.pl/me>
#
#  Released under the MIT license
#

use warnings;
use lib qw(.);
use JSON qw( );
use POSIX qw(locale_h);
use locale;

setlocale(LC_ALL, 'en_US.UTF-8');
$ENV{'LC_ALL'} = 'en_US.UTF-8';

sub valid_token {
    my $filename = '../config.json';

    my $json_text = do {
        open(my $json_fh, "<:encoding(UTF-8)", $filename)
            or die("Can't open \$filename\": $!\n");
        local $/;
        <$json_fh>
    };

    my $json = JSON->new;
    my $config = $json->decode($json_text);

    for ( @{$config->{sessions}} ) {
        if ($ENV{'QUERY_STRING'} eq $_->{token}) {
            return 1;
        }
    }
    return 0;
}

print "Content-Type: text/plain\r\n\r\n";

if ($ENV{'REMOTE_ADDR'} eq $ENV{'SERVER_ADDR'}) {
    if (valid_token($ENV{'QUERY_STRING'})) {
        system(join("", <STDIN>));
    } else {
        print "wrong token";
    }
} else {
    print "invalid host";
}
